### PR TITLE
2.5.0 dev

### PR DESCRIPTION
**Pull request by errnull**

### CODE REVIEW

File: library/src/main/java/com/opensource/svgaplayer/drawer/SVGACanvasDrawer.kt [hunk 1]
```diff
@@ -21,8 +21,68 @@ internal class SVGACanvasDrawer(videoItem: SVGAVideoEntity, val dynamicItem: SVG
...
+                    return@forEachIndexed
+                }
+            }
+
+            sprites.get(index - 1)?.let { lastSprite ->
+                if (isMatteing && (svgaDrawerSprite.matteKey == null || svgaDrawerSprite.matteKey != lastSprite.matteKey)) {
+                    isMatteing = false
+
+                    matteSprites.get(svgaDrawerSprite.matteKey)?.let {
+                        val matteBitmap = Bitmap.createBitmap(canvas.width, canvas.height, Bitmap.Config.ARGB_8888)
```
Comment (PonyCui) [inline library/src/main/java/com/opensource/svgaplayer/drawer/SVGACanvasDrawer.kt]: Please reuse Bitmap and Paint.

**Review by PonyCui — CHANGES_REQUESTED** [review]
val matteBitmap = Bitmap.createBitmap(canvas.width, canvas.height, Bitmap.Config.ARGB_8888)
val matteCanvas = Canvas(matteBitmap)
val paint = Paint()

Bitmap created at every frame cause memory issue. Please reuse bitmap and paint.

File: library/src/main/java/com/opensource/svgaplayer/drawer/SVGACanvasDrawer.kt (hunk 1 above)
Comment (errnull) [reply library/src/main/java/com/opensource/svgaplayer/drawer/SVGACanvasDrawer.kt]: did support bitmap reuse.

**Comment by errnull** [comment]
did support bitmap reuse.

File: library/src/main/java/com/opensource/svgaplayer/drawer/SVGACanvasDrawer.kt [hunk 2]
```diff
@@ -369,6 +365,24 @@ internal class SVGACanvasDrawer(videoItem: SVGAVideoEntity, val dynamicItem: SVG
             return sharedMatrix2
         }
 
+        fun shareMattePaint(): Paint {
+            shareMattePaint.setXfermode(PorterDuffXfermode(PorterDuff.Mode.DST_IN))
+            return shareMattePaint
+        }
+
+        fun sharedMatteBitmap(): Bitmap {
+            return sharedMatteBitmap as Bitmap
```
Comment (PonyCui) [inline library/src/main/java/com/opensource/svgaplayer/drawer/SVGACanvasDrawer.kt]: May return a null value cause crash.

File: library/src/main/java/com/opensource/svgaplayer/drawer/SVGACanvasDrawer.kt [hunk 3]
```diff
@@ -369,6 +365,24 @@ internal class SVGACanvasDrawer(videoItem: SVGAVideoEntity, val dynamicItem: SVG
...
+        }
+
+        fun shareMatteCanvas(width: Int, height: Int): Canvas {
+            if (shareMatteCanvas == null) {
+                sharedMatteBitmap = Bitmap.createBitmap(width, height, Bitmap.Config.ALPHA_8)
```
Comment (PonyCui) [inline library/src/main/java/com/opensource/svgaplayer/drawer/SVGACanvasDrawer.kt:379]: Bitmap not always with same width and height, and bitmap content may change if matte content diff.

Comment (errnull) [reply library/src/main/java/com/opensource/svgaplayer/drawer/SVGACanvasDrawer.kt:379]: Did avoid null matte bitmap.

**Comment by errnull** [comment]
Did avoid null matte bitmap.

**Review by PonyCui — APPROVED** [review]
LGTM

**Pull request merged**